C front-end: fix promotion order for types ranking lower than int

tautschnig opened the pull request:
The ranks specified in the enum are only correct if sizeof(int) > sizeof(short)
and sizeof(int) > sizeof(char). The code was supposed to fix this up, but none
of the branches was feasible as max_type would always be at least INT.

Fixes: #1748

tautschnig committed:
C front-end: fix promotion order for types ranking lower than int

The ranks specified in the enum are only correct if sizeof(int) > sizeof(short)
and sizeof(int) > sizeof(char). The code was supposed to fix this up, but none
of the branches was feasible as max_type would always be at least INT.

Fixes: #1748

peterschrammel reviewed (approved):
LGTM. Is it possible to add the tests posted in #1748?

tautschnig commented:
> Is it possible to add the tests posted in #1748?

My apologies, I really need to work on my use of `git add` ... -- the test had all been prepared alongside the fix.

tautschnig committed:
Support for --16 on Visual Studio, no _WIN64 in 32-bit mode

tautschnig committed:
Do not use non-trivial system headers with --32

Function declarations in system headers would cause type conflicts when running
on 64-bit systems. Just add a proper declaration of malloc in place instead.

tautschnig commented:
@peterschrammel You might also want to re-review as two more commits have been added to address bugs that surfaced.

peterschrammel commented:
@tautschnig, thanks for the tests. clang-format is picky; otherwise ready to merge.

tautschnig commented:
clang-format fixes applied.

kroening reviewed (approved):
Slightly obscure scenarios, but ok!
